### PR TITLE
sessionctx: move shardRand from TransactionContext to SessionVars to reduce allocation

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### What is changed and how it works?

TransactionCtx is reset every time, so the `shardRand` is re-allocated everytime.
https://github.com/pingcap/tidb/blob/2ca37cb156f95d0d70575ab66b886822efa1e5bd/sessionctx/variable/session.go#L249-L252

```
cd session;
go test -run XXX -bench BenchmarkAutoIncrement -cpuprofile cpu.out -benchtime 20s -benchmem
```

```
BenchmarkAutoIncrement-16    	  182841	    128616 ns/op	   40619 B/op	     428 allocs/op
vs
BenchmarkAutoIncrement-16    	  186334	    118309 ns/op	   35175 B/op	     426 allocs/op
```

Before the change:

image

After the change:

image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -246,21 +245,22 @@ type SavepointRecord struct {
 }
 
 // GetCurrentShard returns the shard for the next `count` IDs.
-func (tc *TransactionContext) GetCurrentShard(count int) int64 {
-	if tc.shardRand == nil {
-		tc.shardRand = rand.New(rand.NewSource(int64(tc.StartTS))) // #nosec G404
+func (sv *SessionVars) GetCurrentShard(count int) int64 {
```
Comment (hawkingrei) [inline sessionctx/variable/session.go:248]:
```suggestion
func (s *SessionVars) GetCurrentShard(count int) int64 {
```

make linter happy

Comment (tiancaiamao) [reply sessionctx/variable/session.go]: Done

**Review by hawkingrei — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**